[RFC] Eliminate possible race between ALTER_METADATA and MERGE_PARTS

azat opened the pull request:
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Eliminate possible race between ALTER_METADATA and MERGE_PARTS (that leads to checksum mismatch - CHECKSUM_DOESNT_MATCH)

Cc: @tavplubix 

Refs: https://github.com/ClickHouse/ClickHouse/pull/28432

azat committed:
Add infrastructure for testing replicated MergeTree queue

- replicated_queue_fail_next_entry - to fail next queue entry
- replicated_queue_unfail_entries - to "unfail" all queue entries (if
  any)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Add a test for ALTER_METADATA vs MERGE_PARTS race (CHECKSUM_DOESNT_MATCH)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Eliminate possible race between ALTER_METADATA and MERGE_PARTS

v2: move metadata version check after checking that the part is not covering part
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-ch-test-poll commented:
<!-- automatic status comment for PR #57755 from azat/ClickHouse:rmt/ALTER_METADATA-MERGE_PARTS-race -->
*This is an automated comment for commit 3d05c4b0a11fe377676101b233e08a8fb2652425 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57755/3d05c4b0a11fe377676101b233e08a8fb2652425/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7170660220">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57755/3d05c4b0a11fe377676101b233e08a8fb2652425/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

tavplubix commented on src/Storages/MergeTree/MergeFromLogEntryTask.cpp:
```diff
@@ -109,6 +111,18 @@ ReplicatedMergeMutateTaskBase::PrepareResult MergeFromLogEntryTask::prepare()
             };
         }
 
+        int32_t part_metadata_version = source_part_or_covering->getMetadataVersion();
+        if (part_metadata_version > metadata_version)
+        {
+            LOG_DEBUG(log, "Source part metadata version {} is newer then the table metadata version {}. ALTER_METADATA is still in progress.",
```
Let's move this below just to make sure that it's actually the source part (and not a covering part)

azat replied:
Done.

tavplubix reviewed (commented):
In general LGTM, but the new test is flaky and the failure of `01079_parallel_alter_detach_table_zookeeper` looks suspicious

azat committed:
Fix 02943_rmt_alter_metadata_merge_checksum_mismatch flakiness

Disable keeper injection to make part names static, unlike here [1]:

    azat@s1:~/ch/tmp/57755 [1] {elapsed: 301s}$ zstd -cdq clickhouse-server.log.zst | grep c287beae-b56e-4193-b4c2-812ca5c52919
    2023.12.30 03:20:15.984668 [ 1956 ] {c287beae-b56e-4193-b4c2-812ca5c52919} <Debug> executeQuery: (from [::1]:34846) (comment: 02943_rmt_alter_metadata_merge_checksum_mismatch.sh) insert into data_r2 (key) values  (stage: Complete)
    2023.12.30 03:20:15.987023 [ 1956 ] {c287beae-b56e-4193-b4c2-812ca5c52919} <Trace> ContextAccess (default): Access granted: INSERT(key) ON test_y82swg5w.data_r2
    2023.12.30 03:20:16.007771 [ 1956 ] {c287beae-b56e-4193-b4c2-812ca5c52919} <Trace> test_y82swg5w.data_r2 (fcf801a4-4edd-4209-b52a-4400eb4c4a4c): Trying to reserve 1.00 MiB using storage policy from min volume index 0
    2023.12.30 03:20:16.008769 [ 1956 ] {c287beae-b56e-4193-b4c2-812ca5c52919} <Trace> DiskLocal: Reserved 1.00 MiB on local disk `default`, having unreserved 94.58 GiB.
    2023.12.30 03:20:16.034847 [ 1956 ] {c287beae-b56e-4193-b4c2-812ca5c52919} <Trace> MergedBlockOutputStream: filled checksums all_1_1_0 (state Temporary)
    2023.12.30 03:20:16.037815 [ 1956 ] {c287beae-b56e-4193-b4c2-812ca5c52919} <Debug> test_y82swg5w.data_r2 (fcf801a4-4edd-4209-b52a-4400eb4c4a4c) (Replicated OutputStream): Wrote block with ID 'all_16201685294980115408_4608068419994166055', 1 rows
    2023.12.30 03:20:16.144259 [ 1956 ] {c287beae-b56e-4193-b4c2-812ca5c52919} <Trace> test_y82swg5w.data_r2 (fcf801a4-4edd-4209-b52a-4400eb4c4a4c) (Replicated OutputStream): ZooKeeperWithFaultInjection call FAILED: seed=8123498043031264807 func=tryMulti path=/clickhouse/tables/test_y82swg5w/data/blocks/all_16201685294980115408_4608068419994166055 code=Operation timeout message=Fault injection after operation
    2023.12.30 03:20:16.148988 [ 1956 ] {c287beae-b56e-4193-b4c2-812ca5c52919} <Trace> test_y82swg5w.data_r2 (fcf801a4-4edd-4209-b52a-4400eb4c4a4c) (Replicated OutputStream): ZooKeeperRetriesControl: commitPart: setKeeperError: error=Operation timeout message=Fault injection after operation
    2023.12.30 03:20:16.149232 [ 1956 ] {c287beae-b56e-4193-b4c2-812ca5c52919} <Debug> test_y82swg5w.data_r2 (fcf801a4-4edd-4209-b52a-4400eb4c4a4c) (Replicated OutputStream): ZooKeeperRetriesControl: commitPart: will retry due to error: retry_count=1/100 timeout=1ms error=Operation timeout message=Fault injection after operation
    2023.12.30 03:20:16.155148 [ 1956 ] {c287beae-b56e-4193-b4c2-812ca5c52919} <Trace> test_y82swg5w.data_r2 (fcf801a4-4edd-4209-b52a-4400eb4c4a4c): Renaming temporary part tmp_insert_all_1_1_0 to all_3_3_0 with tid (1, 1, 00000000-0000-0000-0000-000000000000).
    2023.12.30 03:20:16.161514 [ 1956 ] {c287beae-b56e-4193-b4c2-812ca5c52919} <Debug> test_y82swg5w.data_r2 (fcf801a4-4edd-4209-b52a-4400eb4c4a4c) (Replicated OutputStream): ZooKeeperRetriesControl: commitPart: succeeded after: Iterations=2 Total keeper failures=1/100
    2023.12.30 03:20:16.166390 [ 1956 ] {c287beae-b56e-4193-b4c2-812ca5c52919} <Debug> executeQuery: Read 1 rows, 4.00 B in 0.181492 sec., 5.509884733211382 rows/sec., 22.04 B/sec.
    2023.12.30 03:20:16.166950 [ 1956 ] {c287beae-b56e-4193-b4c2-812ca5c52919} <Debug> MemoryTracker: Peak memory usage (for query): 3.13 MiB.
    2023.12.30 03:20:16.168314 [ 1956 ] {c287beae-b56e-4193-b4c2-812ca5c52919} <Debug> TCPHandler: Processed in 0.186244473 sec.

  [1]: https://s3.amazonaws.com/clickhouse-test-reports/57755/a12df35be4c6954e683dbea53c00599ca6a96d5d/stateless_tests_flaky_check__asan_.html

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
>In general LGTM, but the new test is flaky

Fixed.

> and the failure of 01079_parallel_alter_detach_table_zookeeper looks suspicious

It did not failed this time...

>Stateless tests (release, s3 storage) [1/2] — fail: 1, passed: 2944, skipped: 31

- `02483_test_reverse_dns_resolution`
